Add codegen option instead of derive

amaloz opened the pull request:
This MR exposes the `zero`, `one`, and `multiplicative_generator` values as constants rather than functions. It might be preferred to only have the constants version---and I'm happy to make that change if so---but for now I've left the original function versions as well.

amaloz committed:
Expose the zero, one, and generator elements as constants

amaloz committed:
add comments

tarcieri commented:
Related: #87

marcgalois committed:
Add codegen option instead of derive

doesn't cache the output of the proc macro, and using the derive form of
this code can cause slowdowns in `cargo check` and `rust-analyzer`.

As an alternative, this commit introduces an option to generate the
`PrimeField` implementation in a `build.rs` script, to avoid the
slowdown.

str4d commented:
This PR is now almost entirely the codegen changes, which cannot be reviewed in the same PR as the trait changes. So I'm going to make my own PR on top of #93 that adds the associated constants, and we can use that to investigate the ecosystem and confirm whether it will cause any issues for downstream trait implementors (https://github.com/zkcrypto/ff/issues/87#issuecomment-1294891608).

str4d commented:
#87 has now been closed by #94. This PR now needs rebasing, after which it will only contain the proposed codegen changes.

marcgalois committed:
Upgrade subtle dependency

marcgalois committed:
Be more permissive with subtle dependency

marcgalois committed:
chore: bump rust version

marcgalois committed:
chore: allow more subtle versions

marcgalois committed:
use as_mut_slice() instead of as_ref()

The latter was causing issues for type inference.

ChrisEPhifer committed:
Upgrade to rand_core 0.10.1.

- Bump rust-toolchain to 1.97.0
- Bump edition to 2024
- Bump rand_core to 0.10.1 and rand to 0.10.2
